# 01 - Data Validation & Preparation (SQL-Based)

Data preparation and feature engineering were performed in SQL (SQLite)
This notebook validates the final dataset and ensures it is ready for analysis.

## 1. Load Data

In [1]:
import sqlite3
import pandas as pd

In [ ]:
conn = sqlite3.connect("db/telecom.db")

df = pd.read_sql_query("SELECT * FROM telecom_features", conn)
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,MonthlyCharges,PhoneService,MultipleLines,InternetService,...,Contract,PaperlessBilling,PaymentMethod,TotalCharges,Churn,tenure_group,avg_charge_per_month,services_count,is_auto_pay,is_high_monthly_charge
0,7590-VHVEG,Female,No,Yes,No,1,29.85,No,No phone service,DSL,...,Month-to-month,Yes,Electronic check,29.85,No,0-6,29.850000,1,0,0
1,5575-GNVDE,Male,No,No,No,34,56.95,Yes,No,DSL,...,One year,No,Mailed check,1889.50,No,25+,55.573529,2,0,0
2,3668-QPYBK,Male,No,No,No,2,53.85,Yes,No,DSL,...,Month-to-month,Yes,Mailed check,108.15,Yes,0-6,54.075000,2,0,0
3,7795-CFOCW,Male,No,No,No,45,42.30,No,No phone service,DSL,...,One year,No,Bank transfer (automatic),1840.75,No,25+,40.905556,3,1,0
4,9237-HQITU,Female,No,No,No,2,70.70,Yes,No,Fiber optic,...,Month-to-month,Yes,Electronic check,151.65,Yes,0-6,75.825000,0,0,0


## 2. Dataset Overview

In [3]:
df.shape

(7043, 26)

In [4]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'MonthlyCharges', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod', 'TotalCharges', 'Churn',
       'tenure_group', 'avg_charge_per_month', 'services_count', 'is_auto_pay',
       'is_high_monthly_charge'],
      dtype='object')

- Number of rows: 7,043
- Number of columns: 26
- Target variable: `Churn`

## 3. Data Dictionary
### Column Groups

- **Customer Information:** gender, SeniorCitizen, Partner, Dependents

- **Services:** InternetService, TechSupport, StreamingTV, etc.

- **Billing:** Contract, MonthlyCharges, PaymentMethod, PaperlessBilling

- **Target:** Churn

## 4. Data Quality Check
Check for missing values and validate numeric distributions.

In [5]:
df.isnull().sum().sort_values(ascending=False).head(5)

avg_charge_per_month    11
TotalCharges            11
customerID               0
gender                   0
is_auto_pay              0
dtype: int64

In [6]:
df.describe()

,tenure,MonthlyCharges,TotalCharges,avg_charge_per_month,services_count,is_auto_pay,is_high_monthly_charge
count,7043.000000,7043.000000,7032.000000,7032.000000,7043.000000,7043.000000,7043.000000
mean,32.371149,64.761692,2283.300441,64.799424,2.037910,0.435326,0.380094
std,24.559481,30.090047,2266.771362,30.185891,1.847682,0.495835,0.485444
min,0.000000,18.250000,18.800000,13.775000,0.000000,0.000000,0.000000
25%,9.000000,35.500000,401.450000,36.179891,0.000000,0.000000,0.000000
50%,29.000000,70.350000,1397.475000,70.373239,2.000000,0.000000,0.000000
75%,55.000000,89.850000,3794.737500,90.179560,3.000000,1.000000,1.000000
max,72.000000,118.750000,8684.800000,121.400000,6.000000,1.000000,1.000000


## 5. Cleaning & Feature Engineering
Data cleaning and feature engineering were performed in SQL.

Key transformations:

- Converted `TotalCharges` to numeric with null handling  

- Created `tenure_group` to analyze customer lifecycle  

- Created `avg_charge_per_month` for normalized spending 

- Created `services_count` to measure product usage  

- Created `is_auto_pay` from payment method  


## 6. Summary
- SQL used as primary transformation layer
- Monthly Charges used as revenue proxy
- No time-series data available
- Data set is clean and ready for analysis